### DISCUSSION TITLE
flash_firmware --permanent not working

### DISCUSSION THREAD

**Discussion by chaotix-** [Peter-van-Tol/LiteX-CNC#115] [General]
Hello.

I have a problem flashing a firmware with --permanent. Without it everything works fine. But using the same flash command with the same file just adding "--permanent" flashes the FPGA but it does not work (or at least react to network request). This is on a Pi 5 with the FPGA-JTAG-Headers connected to the appropriate GPIO pins on the extension header.

What am I doing wrong?

With --permanent flashing takes a lot longer and the FPGA does not answer ping requests:
```
lars@cncpi:~/linuxcnc/configs/CNC $ time litexcnc flash_firmware cncpi-5a-75e_v8.2/gateware/colorlight_5a_75e.svf --permanent
INFO: Converting /home/lars/linuxcnc/configs/CNC/cncpi-5a-75e_v8.2/gateware/colorlight_5a_75e.svf to flash...
IDCODE in bitstream is 0x41111043
INFO: Conversion complete, output stored in /home/lars/linuxcnc/configs/CNC/cncpi-5a-75e_v8.2/gateware/colorlight_5a_75e.flash!
INFO: Flashing firmware to LED-card.
Open On-Chip Debugger 0.12.0+dev-00761-gd09ff4764 (2024-10-24-21:06)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Info : Linux GPIOD JTAG/SWD bitbang driver
Info : Note: The adapter "linuxgpiod" doesn't support configurable speed
Info : JTAG tap: ecp5.tap tap/device found: 0x41111043 (mfg: 0x021 (Lattice Semi.), part: 0x1111, ver: 0x4)
Warn : gdb services need one or more targets defined
svf processing file: "/home/lars/linuxcnc/configs/CNC/cncpi-5a-75e_v8.2/gateware/colorlight_5a_75e.flash"
95%    INFO: Successfully flashed the LED-card

real	0m48.874s
user	0m1.212s
sys	0m3.358s
lars@cncpi:~/linuxcnc/configs/CNC $ ping 10.0.0.10
PING 10.0.0.10 (10.0.0.10) 56(84) bytes of data.
^C
--- 10.0.0.10 ping statistics ---
2 packets transmitted, 0 received, 100% packet loss, time 1003ms
```
Just dropping the --permanent argument and the file flashes a lot quicker and the FPGA responds to ping packets immediately:
```
lars@cncpi:~/linuxcnc/configs/CNC $ time litexcnc flash_firmware cncpi-5a-75e_v8.2/gateware/colorlight_5a_75e.svf 
INFO: Flashing firmware to LED-card.
Open On-Chip Debugger 0.12.0+dev-00761-gd09ff4764 (2024-10-24-21:06)
Licensed under GNU GPL v2
For bug reports, read
	http://openocd.org/doc/doxygen/bugs.html
Info : Linux GPIOD JTAG/SWD bitbang driver
Info : Note: The adapter "linuxgpiod" doesn't support configurable speed
Info : JTAG tap: ecp5.tap tap/device found: 0x41111043 (mfg: 0x021 (Lattice Semi.), part: 0x1111, ver: 0x4)
Warn : gdb services need one or more targets defined
svf processing file: "/home/lars/linuxcnc/configs/CNC/cncpi-5a-75e_v8.2/gateware/colorlight_5a_75e.svf"
95%    INFO: Successfully flashed the LED-card

real	0m3.344s
user	0m0.551s
sys	0m2.538s
lars@cncpi:~/linuxcnc/configs/CNC $ ping 10.0.0.10
PING 10.0.0.10 (10.0.0.10) 56(84) bytes of data.
64 bytes from 10.0.0.10: icmp_seq=1 ttl=128 time=0.156 ms
^C
--- 10.0.0.10 ping statistics ---
1 packets transmitted, 1 received, 0% packet loss, time 0ms
rtt min/avg/max/mdev = 0.156/0.156/0.156/0.000 ms
```

Thank you for you help,
Lars

**Comment by romanetz** [comment]
Try to full erase flash before writing new firmware. ECP5 FPGA supports dual configuration feature. That means the FPGA will consider your firmware as "broken" and will try to run a backup copy of stock one

**Comment by Peter-van-Tol** [comment]
In my experience it is normal that flashing the NAND takes longer.

Maybe something has changed in the design of the card (for example the i5 has also some problems). 

This weekend I'm away on holiday. Monday I can look into your problem. To reproduce your settings, can you share:
‐ your config
- version of LitexCNC (from PyPi or build from GitHub)
- a picture od the card, to identify a possible new version

**Comment by chaotix-** [comment]
@romanetz : I don't know how to do this. I looked at the documentation for OpenOCD and this didn't really help me as I don't have any experience with FPGAs and the instructions for erasing the flash looked complicated and I don't know if I can mess something up if I mix up NAND/NOR and the regions.

@Peter-van-Tol : Thanks for looking into this (and you fantastic work on this project). Enjoy your holiday. This is absolutely not time-sensitive. It works by flashing it after every restart. And I still need to solder my BOBs (inspired by your designs. I wanted differential drives with AM26LS31 for the motors as on your Pi-HAT but boards in Raspberry Pi size that I can clip onto DIN-rails).

I installed LitexCNC via pip.

```
$ pip show litexcnc
Name: litexcnc
Version: 1.2.4
```

I also attached pictures of the board. They look identical to me to the V6.0 pictures at https://github.com/q3k/chubby75/blob/master/5a-75e/hardware_V6.0.md.

![colorlight-5a-75e-v8 2-front](https://github.com/user-attachments/assets/84d7ad2b-ba48-4196-aa32-7b1439c494bd)
![colorlight-5a-75e-v8 2-back](https://github.com/user-attachments/assets/f80f2bff-92b9-492b-82d2-b89d8309008b)
